### PR TITLE
perf: improve tuple validation by precomputing TTUs

### PR DESCRIPTION
**Pull request by miparnisari**
## Description
Closes https://github.com/openfga/openfga/issues/1104

### CODE REVIEW

File: pkg/typesystem/typesystem.go [hunk 1]
```diff
@@ -177,6 +182,8 @@ func New(model *openfgav1.AuthorizationModel) *TypeSystem {
 			}
 
 			tdRelations[relation] = r
+			ttus := make([]*openfgav1.TupleToUserset, 0)
```
Comment (adriantam) [inline pkg/typesystem/typesystem.go:185]: Are we able to know the size and preallocate instead of allocating per slice?

Comment (miparnisari) [reply pkg/typesystem/typesystem.go:185]: Hmmm we'd have to recursively walk through the relation definition, which we do in the next line. I could preallocate a fixed size like 100, but i don't see much benefit. The problem with not preallocating is that if you add many items, Go will have to copy all elements into bigger and bigger arrays, but here, we won't have many elements. And if we do, we only pay the penalty once when building the model

Comment (adriantam) [reply pkg/typesystem/typesystem.go:185]: If we don't have many items, that's fine.  I was worried that if we have lots of items, there is a lot of copying / moving elements into larger array.

**Review by adriantam — DISMISSED** [review]
Are we able to measure the perf. difference between the two implementations?

File: pkg/typesystem/typesystem.go (hunk 1 above)
Comment (miparnisari) [reply pkg/typesystem/typesystem.go:185]: I mean, we don't have a limit of number of statements in a relation definition, but I can't imagine someone writing a very long line? `define viewer as a1 or a2 or a3 .... or a100`?

**Comment by miparnisari** [comment]
@adriantam it's in the Pr summary

image

The last column is the one that matters, less than 1 means improvement over `main`, more than 1 means regression.

🤔 I need a new benchmark that runs lots of tuple validations, which is where I expect this PR would show an improvement...

**Comment by miparnisari** [comment]
@adriantam  it looks like double as fast based on number of iterations made :D 

```
On main:

[24/11/23 11:43:47] ~/GitHub/openfga (main) $ go test ./internal/validation -bench=BenchmarkValidateTuple  -cpu 1 -benchmem -benchtime 5s -timeout 0 -run=^a
goos: darwin
goarch: amd64
pkg: github.com/openfga/openfga/internal/validation
cpu: Intel(R) Core(TM) i5-1038NG7 CPU @ 2.00GHz
BenchmarkValidateTuple   1292979              4639 ns/op            1232 B/op         11 allocs/op
PASS
ok      github.com/openfga/openfga/internal/validation  11.159s
[24/11/23 11:45:04] ~/GitHub/openfga (main) $ 


This PR:

[24/11/23 11:45:22] ~/GitHub/openfga (perf-improv) $  go test ./internal/validation -bench=BenchmarkValidateTuple  -cpu 1 -benchmem -benchtime 5s -timeout 0 -run=^a
goos: darwin
goarch: amd64
pkg: github.com/openfga/openfga/internal/validation
cpu: Intel(R) Core(TM) i5-1038NG7 CPU @ 2.00GHz
BenchmarkValidateTuple   2022118              3015 ns/op             120 B/op          3 allocs/op
PASS
ok      github.com/openfga/openfga/internal/validation  9.404s

```

**Review by adriantam — APPROVED** [review]

**Pull request merged**